### PR TITLE
feat: support to fetch `route_id` and `service_id` via `ctx.var`

### LINKED ISSUE #2275: feat: support to fetch `consumer_name`, `route_id` and `service_id` via `ctx.var`
here is the fake code:

```lua
local consumer_name = ctx.var.consumer_name
local route_id = ctx.var.route_id
local service_id = ctx.var.service_id
```

This makes it easier for the plugin code to get the values of these variables.

### PR DESCRIPTION
**Pull request by Firstsawyou**
### What this PR does / why we need it:
fix #2275 
Add support to fetch `consumer_name` via `ctx.var`,  the way of `consumer_name` is added by the next `PR`, because waiting for #2270  PR merge.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: t/core/ctx.t [hunk 1]
```diff
@@ -155,3 +155,183 @@ GET /t?a=aaa
...
+
+
+
+=== TEST 8: route_id
+--- request
+GET /hello
+--- response_body
+hello world
+--- error_log
+route_id: 1
```
Comment (membphis) [inline t/core/ctx.t:229]: need to add a checkpoint:

```
--- no_error_log
[error]

File: t/core/ctx.t [hunk 2]
```diff
@@ -155,3 +155,183 @@ GET /t?a=aaa
 --- error_code: 500
 --- error_log
 invalid argument, expect string value
+
+
+
+=== TEST 7: add route
```
Comment (membphis) [inline t/core/ctx.t:161]: We should use a meaningful title

File: t/core/ctx.t [hunk 3]
```diff
@@ -155,3 +155,183 @@ GET /t?a=aaa
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: route_id
```
Comment (membphis) [inline t/core/ctx.t:223]: We should use a meaningful title

File: t/core/ctx.t [hunk 4]
```diff
@@ -155,3 +155,183 @@ GET /t?a=aaa
...
+--- request
+GET /hello
+--- response_body
+hello world
+--- error_log
+route_id: 1
+
+
+
+=== TEST 9: create service (id:1)
```
Comment (membphis) [inline t/core/ctx.t:233]: We should use a meaningful title

File: t/core/ctx.t [hunk 5]
```diff
@@ -155,3 +155,183 @@ GET /t?a=aaa
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 11: service_id
```
Comment (membphis) [inline t/core/ctx.t:331]: We should use a meaningful title

**Review by membphis — CHANGES_REQUESTED** [review]
and we need a test case for `server_id` is `nil`. eg: the `route` object not bind any service object.

File: t/core/ctx.t (hunk 1 above)
Comment (Firstsawyou) [reply t/core/ctx.t:229]: got it.

File: t/core/ctx.t (hunk 2 above)
Comment (Firstsawyou) [reply t/core/ctx.t]: fixed

File: t/core/ctx.t (hunk 3 above)
Comment (Firstsawyou) [reply t/core/ctx.t]: fixed

File: t/core/ctx.t (hunk 4 above)
Comment (Firstsawyou) [reply t/core/ctx.t]: fixed

File: t/core/ctx.t (hunk 5 above)
Comment (Firstsawyou) [reply t/core/ctx.t]: fixed

File: t/core/ctx.t [hunk 6]
```diff
@@ -155,3 +155,264 @@ GET /t?a=aaa
...
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "methods": ["GET"],
+                    "plugins": {
+                        "serverless-pre-function": {
+                            "phase": "access",
+                            "functions" : ["return function() if ngx.ctx.api_ctx.service_id then ngx.log(ngx.ERR, \"service_id: \", ngx.ctx.api_ctx.service_id) end end"]
```
Comment (membphis) [inline t/core/ctx.t:288]: `ngx.ctx.api_ctx.service_id` -> `ngx.ctx.api_ctx.var["service_id"]` ?

Comment (membphis) [reply t/core/ctx.t:288]: we should print the `service_id` always, and check the output value in error log.

`ngx.log(ngx.INFO, "service_id: ", ngx.ctx.api_ctx.var["service_id"] or "empty id")`

**Review by membphis — CHANGES_REQUESTED** [review]

File: t/core/ctx.t [hunk 7]
```diff
@@ -155,3 +155,264 @@ GET /t?a=aaa
...
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                ngx.HTTP_PUT,
+                [[{
+                    "methods": ["GET"],
+                    "plugins": {
+                        "serverless-pre-function": {
+                            "phase": "access",
+                            "functions" : ["return function() ngx.log(ngx.ERR, \"route_id: \", ngx.ctx.api_ctx.route_id) end"]
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/core/ctx.t:173]: got it

File: t/core/ctx.t (hunk 6 above)
Comment (Firstsawyou) [reply t/core/ctx.t:288]: thank you, I got it.

File: t/core/ctx.t [hunk 8]
```diff
@@ -155,3 +155,264 @@ GET /t?a=aaa
...
+
+
+
+=== TEST 14: service_id is not empty
+--- request
+GET /hello
+--- response_body
+hello world
+--- error_log
+service_id: 1
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/core/ctx.t:418]: The title here is not clear. I will revise it later.

File: t/core/ctx.t (hunk 1 above)
Comment (Firstsawyou) [reply t/core/ctx.t:232]: fixed

File: apisix/core/ctx.lua [hunk 9]
```diff
@@ -89,6 +90,14 @@ do
                 key = re_gsub(key, "-", "_", "jo")
                 val = get_var(key, t._request)
 
+            elseif key == "route_id" then
+                local route_id = ngx.ctx.api_ctx.route_id
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply apisix/core/ctx.lua]: fixed

File: apisix/core/ctx.lua [hunk 10]
```diff
@@ -89,6 +90,14 @@ do
...
+                val = route_id
+
+            elseif key == "service_id" then
+                local service_id = ngx.ctx.api_ctx.service_id
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply apisix/core/ctx.lua]: fixed

File: t/core/ctx.t [hunk 11]
```diff
@@ -155,3 +155,264 @@ GET /t?a=aaa
...
+
+
+
+=== TEST 12: service_id is empty
+--- request
+GET /hello
+--- response_body
+hello world
+--- no_error_log
+[error]
```
[the remark this answers is not in the document]
Comment (Firstsawyou) [reply t/core/ctx.t:342]: fixed

**Review by membphis — APPROVED** [review]

**Review by moonming — APPROVED** [review]
